### PR TITLE
Threaded Comments - Delete Comment Test Fixed

### PR DESCRIPTION
**Pull request by gordonbanderson**
@wilr I've fixed the code that the delete comment test was rightly failing.  Locally I am seeing a couple of the HTML sanitize tests failing, which I do not recollect seeing last night when my previous pull request failed against automated testing.  As such submitting another pull request to see if that is still the case.

Could you point me at relevant documentation how I can set up the automated testing on my own branches first so as not to bug people with potentially unnecessary notifications?  Cheers :)

### CODE REVIEW

File: code/extensions/CommentsExtension.php [hunk 1]
```diff
@@ -130,7 +131,9 @@ public function CommentsForm() {
 			'Parent'					=> $this->owner,
 			'AddCommentForm'			=> $form,
 			'ModeratedSubmitted'		=> $moderatedSubmitted,
-			'Comments'					=> $this->getComments()
+			'Comments'					=> $this->getComments(),
+			'ThreadedComments'          => Commenting::get_config_value($this->ownerBaseClass, 'thread_comments'),
+            'MaxThreadedCommentDepth'   => Commenting::get_config_value($this->ownerBaseClass, 'maximum_thread_comment_depth')
```
Comment (wilr) [inline code/extensions/CommentsExtension.php]: Whitespace.

File: code/controllers/CommentingController.php [hunk 2]
```diff
@@ -411,6 +436,21 @@ public function doPostComment($data, $form) {
 			Session::set('CommentsModerated', 1);
 		}
 
+		// ensure that the parent comment can be replied to, if not throw a 403 forbidden error
+		$threaded = Commenting::get_config_value($class, 'thread_comments');
+        if ($threaded) {
```
Comment (wilr) [inline code/controllers/CommentingController.php]: Whitespace issue

File: code/controllers/CommentingController.php [hunk 3]
```diff
@@ -435,8 +480,62 @@ public function doPostComment($data, $form) {
...
+				'CanPost' 					=> Commenting::can_member_post($this->ownerBaseClass),
+				'RssLink'					=> "CommentingController/rss",
+				//'RssLinkPage'				=> "CommentingController/rss/". $this->ownerBaseClass . '/'.$this->owner->ID,
+				'RssLinkPage'				=> "CommentingController/rss/". $this->ownerBaseClass.'FIXME',
+				'CommentsEnabled' 			=> true,
+				'Parent'					=> $this->dataRecord,
+				'AddCommentForm'			=> $form,
+				'ModeratedSubmitted'		=> $moderatedSubmitted,
+				//'ThreadedComments'          => true, // FIXME - hardcoded
+	           // 'MaxThreadedCommentDepth'   => 10
```
Comment (wilr) [inline code/controllers/CommentingController.php]: Whitespace issues

File: templates/CommentsInterface.ss [hunk 4]
```diff
@@ -1,29 +1,18 @@
 <% if CommentsEnabled %>
+	<%	require javascript('comments/javascript/commenting.js') %>
+	<%  require css('comments/css/comments.css') %>
```
Comment (wilr) [inline templates/CommentsInterface.ss]: Should use themedCSS to allow themes to easily override.

File: code/dataobjects/Comment.php [hunk 5]
```diff
@@ -56,13 +66,37 @@ class Comment extends DataObject {
...
+			$this->write();
 		}
+
+		
+	}
+
+
+	private function paddedNumber($i) {
+		// fixme, use config
+		return str_pad($i, 5, '0', STR_PAD_LEFT);
```
Comment (wilr) [inline code/dataobjects/Comment.php]: Can we change this to use config. 5 seems short.

File: code/dataobjects/Comment.php [hunk 6]
```diff
@@ -333,6 +392,21 @@ public function ApproveLink() {
...
+	 */
+	public function ViewCommentLink() {
+		if($this->canView() && $this->Moderated) {
+			$token = SecurityToken::inst();
+
+			return DBField::create_field("Varchar", Director::absoluteURL($token->addToUrl(sprintf(
+				"CommentingController/viewcomment/%s", 'COMMENTID'
+			))));
+		} else {
+			return 'lolnope';
```
Comment (wilr) [inline code/dataobjects/Comment.php]: Ah?

File: code/dataobjects/Comment.php [hunk 7]
```diff
@@ -418,4 +492,17 @@ public function Gravatar() {
 
 		return $gravatar;
 	}
+
+	/*
+	Check if this comment can be replied to
+	- check threading is enabled
+	- comment must have been moderated
+	*/
+	public function CanReply() {
```
Comment (wilr) [inline code/dataobjects/Comment.php]: Can we get proper PHPDoc used here.

File: css/comments.css [hunk 8]
```diff
@@ -0,0 +1,165 @@
+
+
+.typography ul.action-links, .typography ul.comments-list, .typography ul.no-bullet {
```
Comment (wilr) [inline css/comments.css]: The use of .typography is somewhat specific to the theme. Can we wrap the entire comments template in a div class something like 'comments' add on so we don't break exising themes that use .action-links or anything else generic.

File: javascript/commenting.js [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+		$(document).on('onBeforeSubmitNewCommentForm', function(e) {
+			// before form submission, perhaps add a spinner
+			var submitbutton = $('#Form_CommentsForm_action_doPostComment');
+			submitbutton.val('Posting'); // FIXME i18n
+			submitbutton.attr('disabled',1);
+		});
+
+		$(document).on('onAfterServerValidationFailedNewCommentForm', function(e) {
+			var submitbutton = $('#Form_CommentsForm_action_doPostComment');
+			submitbutton.val('Post'); // FIXME i18n
```
Comment (wilr) [inline javascript/commenting.js]: Can either use data attributes in the HTML for passing these strings or ss.lang. I'd suggest data-attributes could be easiest.

**Comment by wilr** [comment]
Great work @gordonbanderson! Added a few quick tidy up notes. This has 20 commits, is it possible for you to tidy those up via rebase to 3-4 commits that cover the functionality. If not, I can give it a tidy up. Will test the functionality over the weekend and merge it in.

**Comment by tractorcow** [comment]
Hiya, just to ping that I'm also going to test and review this PR. Nice work so far. :)

File: templates/CommentsInterface.ss (hunk 4 above)
Comment (tractorcow) [inline templates/CommentsInterface.ss]: Also <% require %> in a condition doesn't actually work. If you need to conditionally require it it has to go into php.

File: code/dataobjects/Comment.php (hunk 7 above)
Comment (tractorcow) [inline code/dataobjects/Comment.php]: We should also use proper can* convention here, including extension points.

**Comment by tractorcow** [comment]
I'm going to attempt to submit various fixes and cleanup on top of this PR in a separate one. I'll keep you in the loop @wilr  and @gordonbanderson

File: code/Commenting.php [hunk 10]
```diff
@@ -40,6 +40,8 @@ class Commenting {
 		'html_allowed' => false, // allow for sanitized HTML in comments
 		'html_allowed_elements' => array('a', 'img', 'i', 'b'),
```
Comment (tractorcow) [inline code/Commenting.php]: I think this whole config system is antiquated and should be using the current `Config` api. Not a problem with your PR just noting it. :)

File: code/Commenting.php [hunk 11]
```diff
@@ -40,6 +40,8 @@ class Commenting {
...
 		'use_preview' => false, // preview formatted comment (when allowing HTML). Requires include_js=true
+		'thread_comments' => false, // allow replies to existing comments.  Requires include_js=true,
```
Comment (tractorcow) [inline code/Commenting.php]: We really need to develop a non-js compatible reply mechanism as well; Not everyone will want to cater to js-only enabled environments.

File: code/dataobjects/Comment.php [hunk 12]
```diff
@@ -333,6 +392,21 @@ public function ApproveLink() {
 			))));
 		}
 	}
+
+	/**
+	 * @return string
+	 */
+	public function ViewCommentLink() {
```
Comment (tractorcow) [inline code/dataobjects/Comment.php]: Where is this method actually used?

File: templates/CommentsInterface_singlecomment.ss [hunk 13]
```diff
@@ -1,31 +1,37 @@
...
-			<% if HamLink %>
-				<li><a href="$HamLink.ATT" class="ham"><% _t('CommentsInterface_singlecomment_ss.ISNTSPAM','this comment is not spam') %></a></li>
-			<% end_if %>
-			<% if DeleteLink %>
-				<li class="last"><a href="$DeleteLink.ATT" class="delete"><% _t('CommentsInterface_singlecomment_ss.REMCOM','remove this comment') %></a></li>
-			<% end_if %>
-		</ul>
-	<% end_if %>
+<% if $ApproveLink || $SpamLink || $HamLink || $DeleteLink %>
+			<ul class="action-links">
```
Comment (tractorcow) [inline templates/CommentsInterface_singlecomment.ss]: Indentation

File: templates/CommentsInterface_singlecomment.ss [hunk 14]
```diff
@@ -1,31 +1,37 @@
-<div class="comment" id="<% if isPreview %>comment-preview<% else %>$Permalink<% end_if %>">
-<% if $Gravatar %><img class="gravatar" src="$Gravatar" alt="Gravatar for $Name" title="Gravatar for $Name" /><% end_if %>
-	$EscapedComment
-</div>
+<div class="comment" data-comment-id="$ID" id="<% if isPreview %>comment-preview<% else %>$Permalink<% end_if %>">
+<% if MarkedAsDeleted %><% _t('DELETED_BY_ADMIN','This comment was deleted by an administrator') %><% else %>
+<% if $Gravatar %><div class="gravatarContainer"><img class="gravatar" src="$Gravatar" alt="Gravatar for $Name" title="Gravatar for $Name" /></div><% end_if %>
+<div class="actualcomment">
```
Comment (tractorcow) [inline templates/CommentsInterface_singlecomment.ss]: Naming is a bit... on the nose, don't you think? :P Not that I can blame you much.

File: extensions/CommentingControllerTwitterUsernameExtension.php [hunk 15]
```diff
@@ -0,0 +1,11 @@
+<?php 
+
+class CommentingControllerTwitterUsernameExtension extends DataExtension {
```
Comment (tractorcow) [inline extensions/CommentingControllerTwitterUsernameExtension.php]: Shouldn't twitter be a separate module?

**Comment by tractorcow** [comment]
I think this solution is a good one for the specific use case, but I'm concerned that making threaded comments JS-only is going to exclude a lot of other use cases. The core module really needs to support reply functionality at a more basic level, allowing users to customise it to rely on JS on a case by case basis. I'm not sure this is a good candidate for the core module, although a forked comments module for js apps would be fine.

I'm going to look at some other solutions put forward by other developers for comparison. :)

File: templates/CommentsInterface.ss (hunk 4 above)
Comment (gordonbanderson) [inline templates/CommentsInterface.ss]: I didn't know this, thanks

File: code/dataobjects/Comment.php (hunk 5 above)
Comment (gordonbanderson) [inline code/dataobjects/Comment.php]: Agree that it should be configurable.

File: code/dataobjects/Comment.php (hunk 12 above)
Comment (gordonbanderson) [inline code/dataobjects/Comment.php]: Grepping it's not :(

File: extensions/CommentingControllerTwitterUsernameExtension.php (hunk 15 above)
Comment (gordonbanderson) [inline extensions/CommentingControllerTwitterUsernameExtension.php]: Possibly though it does need a check in the template as to whether or not the Twitter username is part of the comment during rendering.  Hence why I included it, but  left it optional to enable it.

**Comment by gordonbanderson** [comment]
With help from @wernerkrauss I've sorted out my dev setup regarding coding conventions, white space being the main issue (and my not reading the coding conventions for a while...).  Something for me to keep an eye on in future so as not to waste others' time as I have in this case.  Apologies.

@tractorcow I agree that a non JS version would be ideal given that SilverStripe is a now commonly used for New Zealand Government websites, I've worked on several in my past and have experienced the issues with likes of file sizes, HTML equivalent of info, graceful degradation etc.  This pull request fails in that regard.

My time has been somewhat fragmented whilst developing the changes to this module (you only need to look at commit dates) but due to a change in personal circumstances now have more coding time.  Apologies for the errors that have crept in as a result of this.


Initial motivation was to create a more 'WordPress' like comments experience out of the box and whilst I'm  happy with the solution from a personal perspective I realize it's not a solution for all.  For an example of what I was aiming at, see the comments section of http://www.richardbarrow.com/2014/04/cycle-track-around-suvarnabhumi-airport/ - @RichardBarrow is the most prominent expat blogger in Thailand where I currently live.

I do feel the configuration setup with this module is holding it back in that 1) It's non standard 2) Lots of options for authors to check (i.e. make a change, test each configuration out).  And also I think 'comments_holder_id' is obsolete, in that if this change any associated JavaScript would break (unless of course the relevant id was passed as a parameter such as in a data attribute, but it just seems like extra hastle).


Regarding JavaScript interaction I've refactored the comment changing functions to trigger events, but in some cases the HTML manipulation is still done within that file, CommentsInterface.js, e.g.

```javascript
/*
/*
Deal with reply to buttons
*/
$(document).on("click",".replyButton button",function(e){
	// get the root node of the comment
	var commentbox = $(this).closest('.comment');

	commentbox.parent().append($('#CommentsFormContainer'));//.parent().parent());
	$(this).parent().addClass('hidden'); // hide reply button

	$(this).parent().parent().find('.cancelReplyButton').removeClass('hidden'); // show cancel reply button
	$('.whenReplying').removeClass('hidden');
	$('.newComment').addClass('hidden');
	var parentCommentID = $(this).attr('data-comment-to-reply-to-id');
	$('#Form_CommentsForm_ParentCommentID').val(parentCommentID);
	$.event.trigger({
		type: "replyToCommentButtonClicked",
		commentid: commentbox.attr('data-comment-id')
	});
});
```

My current thinking is that CommentsInterface.js deals only with the following:
* manipulation of comment data after form submission
* triggering events

HTML rendering/manipulation should be passed off to another JavaScript file, that I think should be configurable depending on desired behavior, currently commenting.js in my pull request.  For example the following scenarios would require different behavior after comments submitted.
* Newest comments first
* Oldest comments first
* Show the first ten with a 'more' button after that a la Facebook.
Happy to hear your opinions regarding the above.

This might mean having instead of comments to install another module that deals with the desired comment ordering and editing behavior.  And also take into account non JavaScript behavior.

@tractorcow You mentioned in one of your comments on https://github.com/silverstripe/silverstripe-comments/pull/94 that the addition of a Twitter username should probably be a separate module.  I kind of agree though it could be argued that the addition of Gravatars (currently a config option) could also be modularised in the same manner.

In order to provide an 'out of the box' option the comments module would have to check for existence of relevant extensions and render accordingly.  Is there an alternative way to do this?
* Installed 'comments gravatar' module - comments module hsa to check for gravatar module and adds one accordingly
* Installed 'comments twiter username' module - comments module has to check for twiter username and renders twitter widget accordingly.


Thank you both for your feedback, I do feel this module needs some love as I see it as weakness in the SilverStripe ecosystem of modules when compared to likes of WordPress (which seems like the default solution to almost everything in Thailand, hence my wish to improve the SIlverStripe experience in this regard).  With a bit of discussion and community feedback it can become something better than it currently is.  Whilst I realise my pull request is not perfect, I feel it's a step in the right direction.

Regards

Gordon

File: css/comments.css (hunk 8 above)
Comment (gordonbanderson) [inline css/comments.css]: I based the CSS on the default Simple theme and in some cases had to use .typography to override.  With modules what is the best practice regarding CSS?

File: javascript/commenting.js (hunk 9 above)
Comment (gordonbanderson) [inline javascript/commenting.js]: Would passing an inline array in JSON format be an option also regarding i18n?  I'm not familiar with approaches to i18n from a JavaScript perspective regarding SilverStripe, advice welcome.

File: code/Commenting.php (hunk 10 above)
Comment (gordonbanderson) [inline code/Commenting.php]: Yeah the one off comments module config system needs ripped out

File: templates/CommentsInterface_singlecomment.ss (hunk 14 above)
Comment (gordonbanderson) [inline templates/CommentsInterface_singlecomment.ss]: Feel free to change :)

File: extensions/CommentingControllerTwitterUsernameExtension.php (hunk 15 above)
Comment (gordonbanderson) [inline extensions/CommentingControllerTwitterUsernameExtension.php]: As per subsequent note below I think the Twitter username case could fall into the same category as having a Gravatar.  My preference is having both available as optional extensions within the comments module, in that the template rendering a comment needs to know if the extensions are enabled or not.  This also helps regarding ripping out some of the old config.

**Comment by tractorcow** [comment]
I've had a bit of a play around with this work and added https://github.com/tractorcow/silverstripe-comments/tree/pr/94. I haven't done much except respond to some of the specific comments here (indentation, etc). Feel free to use it.

**Comment by tractorcow** [comment]
I'm going to close this PR for now since we've merged a threaded comment solution into master.

**Pull request closed without merging**